### ISSUE TITLE
Pusher http compatibility

### ISSUE THREAD

**Issue by jlsjonas** [beyondcode/laravel-websockets#102]
![image](https://user-images.githubusercontent.com/932193/52203165-c5b60780-2870-11e9-81f8-e9c598b8cd1c.png)
It looks like there's some incompatibilities with the pusher/echo client.

Above is trying to make a POST request to the url, however only GET & HEAD is allowed

**Comment by jlsjonas** [comment]
That makes sense (probably should be in the docs though?)

For some reason the websocket connection is indeed not established on the server (works fine locally)
`websockets:serve` command shows no output whatsoever (except for the listening message), firewall is open, SELinux set correctly; tried permissive & firewall off just to be sure; any ideas what might be causing it?
(using apache + let's encrypt, ports & hosts set correctly; ssl config (in `config/websockets.php`) contains the local_cert)

**Comment by stayallive** [comment]
Yeah that should be in the docs (considering the many issues about it...).

That can have a million reasons really :)

Make sure you are using the certificate files (https://github.com/beyondcode/laravel-websockets/issues/92#issuecomment-455192084) this bit a few people already and also you have both the local cert and pk config option set.

If certificate is all good check if there are any message in the developer console of the browser that might give some insight in could be breaking.

If there is no output at all in your browser console make sure you are not using a port that is not allowed by the browser (more about that here: https://alex.bouma.me/installing-laravel-websockets-on-forge/#a-note-about-ports).

**Comment by jlsjonas** [comment]
![image](https://user-images.githubusercontent.com/932193/52216845-7d5e1000-2897-11e9-93ea-8da00f730c80.png)

looks like my firewall was really acting up, opening port 6001 & disabling the firewall didn't seem to have been applied. so it's no longer timing out (thanks good old telnet for discovering that 😅 )
However, still no errors whatsoever, but it's cancelling out instead 😅

Note: as webmin generates the let's encrypt certificates, I don't have a "fullchain" as per the linked comment; however Let's Encrypt should be a recognized CA so I'm not sure if related. (and without logs/errors it's very hard to debug)

**Comment by stayallive** [comment]
Looking at this it does still look like a certificate error: https://stackoverflow.com/questions/31664366/chrome-failing-to-connect-to-websocket-server-opcode-1-handshake-was-cancele. I don't think you need to do anything the accepted answer says, but do need to make sure the certificate is correct.

Maybe you could try to create your own `fullchain.pem` combining the contents of `letsencrypt-cert.pem`, `letsencrypt-ca.pem` (in that order), however it does look like `/etc/letsencrypt/live/` could contain your domain name as folder and could have the fullchain.pem file inside of there looking at webmin threads.

There is currently no error if you supply an invalid or incomplete certificate (https://github.com/beyondcode/laravel-websockets/issues/92#issuecomment-455564110) so that definitely doesn't help, but of a hit and miss right now :)

**Comment by stayallive** [comment]
Running the concatenationin a cron should help with that (you need to restart your websocket server each the the certificate changes anyhow), but this is more a problem with webmin than this package I think (except the error reporting for an incorrectly configured certificate ofcourse).

**Comment by jlsjonas** [comment]
The concatenated version ended up with a `failed: Error in connection establishment: net::ERR_SSL_VERSION_OR_CIPHER_MISMATCH`
and `SSL_ERROR_NO_CYPHER_OVERLAP` on firefox

further investigations lead me... nowhere 😅 
I'll create a cert using certbot for l-w I guess.

Either way, I do believe having an option to include the intermediate certificates as a parameter instead of requiring to provide a combined file would be a more lenient solution; there's several CA's that don't provide a combined file, similar to what apache requires

![image](https://user-images.githubusercontent.com/932193/52261325-594b0f00-2929-11e9-98f8-816dcd00196b.png)

Above is a working configuration, to combine them I tried
`cat letsencrypt-cert.pem letsencrypt-ca.pem > letsencrypt-combined.pem`
along with the pk defined separately and  
`cat letsencrypt-key.pem letsencrypt-cert.pem letsencrypt-ca.pem > letsencrypt-combined.pem`
with no success

Thanks for the help so far, the issue might be broader than this package indeed; though I do believe if the configuration works for apache it should arguably work out for this too.

**Comment by stayallive** [comment]
Now I'm all out of ideas too, great problem! If you do find something it would be really useful to leave here or PR a documentation change to include your findings!

> Thanks for the help so far, the issue might be broader than this package indeed; though I do believe if the configuration works for apache it should arguably work out for this too.

Well the current config for this package works for NGINX, so I guess arguments could be made for every server software that handles SSL just a bit differently (don't even get me started about IIS).

As a general note, I would always place your websockets server behind a reverse proxy like NGINX (or I believe Apache can do the same) so the websockets server doesn't have to handle SSL at all and offload those CPU cycles to more competent and faster applications designed for that work, this way you also don't have to worry about restarting your websockets server when a certificate or config changes but restarting the proxy server (be it Apache or NGINX) is enough. This also has the benefit if your app being able to speak in plain text to the websockets server making it slightly faster. For inspiration: I wrote [this blog](https://alex.bouma.me/installing-laravel-websockets-on-forge/) for Forge & NGINX, that might give you some ideas on how to replicate it for Apache if you want to go that route.

**Comment by amirreza1751** [comment]
Hello @jlsjonas .
I have this problem with laravel websockets and it has taken my time. I still get this error on web browser and I don't know what must be changed. can you help me fix this problem by explaining more about your solution?
Here is my error.

**Comment by amirreza1751** [comment]
`WebSocket connection to 'wss://domain:6001/app/ABCDEFG?protocol=7&client=js&version=4.3.1&flash=false' failed: Error in connection establishment: net::ERR_SSL_VERSION_OR_CIPHER_MISMATCH`


I will send my configs if you need.
